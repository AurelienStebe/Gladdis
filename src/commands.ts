import path from 'path'
import fs from 'fs-extra'
import { JSDOM } from 'jsdom'
import fetch from 'node-fetch'
import TurndownService from 'turndown'
import { getDocument } from 'pdfjs-dist'
import { gfm } from 'turndown-plugin-gfm'
import { arrayBuffer } from 'stream/consumers'

import { doGladdis } from './gladdis.js'
import { transcribe } from './utils/whisper.js'
import { parseLinks } from './utils/scanner.js'
import { webBrowser } from './utils/browser.js'
import { getTokenModal } from './utils/loggers.js'
import { loadContext, loadContent } from './utils/loaders.js'

import type { ReadStream } from 'fs'
import type { PDFDocumentProxy } from 'pdfjs-dist'
import type { Context, DiskInterface } from './types/context.js'

export { stringify as stringifyYaml, parse as parseYaml } from 'yaml'

export function parseDom(html: string): Document {
    return new JSDOM(html).window.document
}

export async function getHtml(url: string): Promise<string> {
    return await (await fetch(url)).text()
}

export function turndown(html: string): string {
    const turndown = new TurndownService({
        hr: '---',
        headingStyle: 'atx',
        bulletListMarker: '-',
        codeBlockStyle: 'fenced',
    })

    turndown.use(gfm)

    return turndown.turndown(html)
}

export async function getPdfDoc(stream: ReadStream | Promise<File>): Promise<PDFDocumentProxy> {
    const file = stream instanceof Promise ? (await stream).arrayBuffer() : arrayBuffer(stream)

    return await getDocument(await file).promise
}

export const diskInterface: DiskInterface = {
    readFile: async (p: string) => await fs.readFile(p, 'utf-8'),
    readBinary: async (p: string) => await arrayBuffer(fs.createReadStream(p)),
    appendFile: async (p: string, d: string) => await fs.appendFile(p, d),
    deleteFile: async (p: string) => await fs.remove(p),
    pathExists: async (p: string) => await fs.pathExists(p),
    pathEnsure: async (p: string) => await fs.ensureDir(p),
    baseName: (p: string, e?: string) => path.basename(p, e),
    extName: (p: string) => path.extname(p),
    joinPath: (...p: string[]) => path.join(...p),
    dirName: (p: string) => path.dirname(p),
}

export async function chatWithGladdis(context: Context): Promise<void> {
    await doGladdis(loadContent(await loadContext(prepareContext(context))))
}

export async function processContent(context: Context): Promise<void> {
    context = loadContent(await loadContext(prepareContext(context)))

    context.whisper.echoOutput = true
    context.whisper.deleteFile = false

    await Promise.all(
        context.user.history.map(async (message) => {
            message.content = await parseLinks(message.content, context)
        }),
    )

    context.user.prompt = await transcribe(context.user.prompt, context)
    context.user.prompt = await parseLinks(context.user.prompt, context)
    context.user.prompt = await webBrowser(context.user.prompt, context)

    context.user.history.push({
        role: 'user',
        content: context.user.prompt,
        name: context.user.label,
    })

    await context.file.disk.appendFile(context.file.path, getTokenModal(context))
}

export async function processPrompt(context: Context): Promise<void> {
    context = await loadContext(prepareContext(context))
    context.file.text = context.file.text.split('\n---').at(-1) ?? ''
    context = loadContent(context)

    context.whisper.echoOutput = true
    context.whisper.deleteFile = false

    context.user.prompt = await transcribe(context.user.prompt, context)
    context.user.prompt = await parseLinks(context.user.prompt, context)
    context.user.prompt = await webBrowser(context.user.prompt, context)

    context.user.history = [
        {
            role: 'user',
            content: context.user.prompt,
            name: context.user.label,
        },
    ]

    await context.file.disk.appendFile(context.file.path, getTokenModal(context))
}

export function prepareContext(context: Context): Context {
    context.user = context.user ?? {}

    context.user.env = context.user.env ?? process.env
    context.file.disk = context.file.disk ?? diskInterface

    return context
}
