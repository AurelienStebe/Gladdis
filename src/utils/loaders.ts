import { deepmerge } from 'deepmerge-ts'

import { parseYaml } from '../commands.js'
import { parseHistory } from './history.js'
import { writeMissedModal } from './loggers.js'

import type { Context } from '../types/context.js'

const linkRegex = /(?<!<%.*)(?<!!)(\[\[[^\]]+?\]\])(?!.*%>)/gs

export async function loadContext(context: Context): Promise<Context> {
    const callContext = deepmerge(context, { file: { date: new Date() } })
    const fileContext = deepmerge(await loadMarkdown(context), callContext)

    const coreContext = {
        file: {
            name: context.file.disk.baseName(context.file.path, '.md'),
            text: '',
        },
        user: {
            data: context.user.env.GLADDIS_DATA_PATH ?? '__DATA__',
            label: context.user.env.GLADDIS_DEFAULT_USER ?? 'User',
            prompt: '',
            history: [],
        },
        gladdis: {
            label: context.user.env.GLADDIS_NAME_LABEL ?? 'Gladdis',
            config: context.user.env.GLADDIS_CONFIG_FILE,
            model: context.user.env.GLADDIS_DEFAULT_MODEL ?? 'gpt-3.5-turbo',
            temperature: Number(context.user.env.GLADDIS_TEMPERATURE ?? 0),
            top_p_param: Number(context.user.env.GLADDIS_TOP_P_PARAM ?? 100),
            freq_penalty: Number(context.user.env.GLADDIS_FREQ_PENALTY ?? 0),
            pres_penalty: Number(context.user.env.GLADDIS_PRES_PENALTY ?? 0),
        },
        whisper: {
            input: context.user.env.GLADDIS_WHISPER_INPUT ?? 'Gladdis',
            config: context.user.env.GLADDIS_WHISPER_CONFIG,
            model: context.user.env.GLADDIS_WHISPER_MODEL ?? 'whisper-1',
            liveSuffix: context.user.env.GLADDIS_WHISPER_LIVE_SUFFIX ?? 'dictated, but not read',
            readSuffix: context.user.env.GLADDIS_WHISPER_READ_SUFFIX ?? 'transcribed and read',
            temperature: Number(context.user.env.GLADDIS_WHISPER_TEMPERATURE ?? 0),
            echoOutput: (context.user.env.GLADDIS_WHISPER_ECHO_OUTPUT ?? 'true').toLowerCase() === 'true',
            deleteFile: (context.user.env.GLADDIS_WHISPER_DELETE_FILE ?? 'false').toLowerCase() === 'true',
            language: context.user.env.GLADDIS_WHISPER_LANGUAGE_ID,
        },
    }

    const fullContext = deepmerge(coreContext, fileContext) as unknown as Context
    return deepmerge(await loadAIConfig(fullContext), fileContext) as unknown as Context
}

export async function loadAIConfig(context: Context): Promise<Context> {
    const disk = context.file.disk
    let configContext: any = { whisper: {} }

    if (context.gladdis.config !== undefined && context.gladdis.config !== '') {
        if (!context.gladdis.config.toLowerCase().endsWith('.md')) context.gladdis.config += '.md'
        const configPath = disk.joinPath(context.user.data, 'configs', context.gladdis.config)

        if (await disk.pathExists(configPath)) {
            configContext = deepmerge(configContext, {
                file: { path: configPath, disk },
                user: { label: 'System' },
            })

            configContext = await loadMarkdown(configContext as Context)
            const configHistory = deepmerge(context, configContext) as Context

            delete configContext.file
            delete configContext.user

            context = deepmerge(context, configContext)
            context.user.history.unshift(...parseHistory(configHistory))
        } else {
            await writeMissedModal(configPath, 'Config File Not Found', context)
        }
    }

    if (context.whisper.config !== undefined && context.whisper.config !== '') {
        if (!context.whisper.config.toLowerCase().endsWith('.md')) context.whisper.config += '.md'
        const whisperPath = disk.joinPath(context.user.data, 'configs', context.whisper.config)

        if (await disk.pathExists(whisperPath)) {
            let whisperContext: any = {
                file: { path: whisperPath, disk },
            }

            whisperContext = await loadMarkdown(whisperContext as Context)
            configContext.whisper.input = whisperContext.file.text

            whisperContext = deepmerge(configContext, whisperContext)
            context.whisper = deepmerge(context.whisper, whisperContext.whisper)
        } else {
            await writeMissedModal(whisperPath, 'Whisper File Not Found', context)
        }
    }

    return context
}

export async function loadMarkdown(context: Context): Promise<Context> {
    const content = (await context.file.disk.readFile(context.file.path)).trim()

    if (content.startsWith('---\n')) {
        const [frontMatter, ...bodyContent] = content.slice(4).split('\n---')

        context = deepmerge(parseYaml(frontMatter), context)
        context.file.text = bodyContent.join('\n---').trim()
    } else context.file.text = content

    return context
}

export function loadContent(context: Context): Context {
    context.user.history.push(...parseHistory(context))

    for (const message of context.user.history) {
        if (message.role === 'system') {
            message.content = message.content.replace(linkRegex, '!$1')
        }
    }

    if (context.user.history.at(-1)?.role === 'user') {
        context.user.label = context.user.history.at(-1)?.name ?? 'User'
        const content = context.user.history.pop()?.content ?? ''

        context.user.prompt = content.replace(linkRegex, '!$1')
    } else context.user.prompt = ''

    return context
}
