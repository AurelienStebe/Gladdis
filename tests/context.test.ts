import fs from 'fs-extra'

import { describe, it, expect, afterEach } from 'vitest'
import { getLocalEnv, cleanupTest, getTestPath } from './tests-utils.js'

import { diskInterface } from '../src/commands.js'
import { loadContext } from '../src/utils/loaders.js'

describe('loadContext() function', () => {
    afterEach(async () => await cleanupTest('__DATA__', 'input.md'))

    it('loads the default values', async () => {
        const filePath = getTestPath('input.md')
        await fs.ensureFile(filePath)

        const callContext = { file: { path: filePath } } as any
        callContext.user = { env: getLocalEnv('__DATA__') }
        callContext.file.disk = diskInterface
        const context = await loadContext(callContext)

        expect(context.file.path).toBe(filePath)
        expect(context.user.data).toBe(getTestPath('__DATA__'))

        expect(context).toMatchSnapshot({
            file: {
                date: expect.any(Date),
                name: 'input',
                path: expect.any(String),
                text: '',
                disk: expect.any(Object),
            },
            user: {
                env: expect.any(Object),
                data: expect.any(String),
                label: 'User',
                prompt: '',
                history: [],
            },
            gladdis: {
                label: 'Gladdis',
                config: undefined,
                model: 'gpt-3.5-turbo',
                temperature: 0,
                top_p_param: 100,
                freq_penalty: 0,
                pres_penalty: 0,
            },
            whisper: {
                input: 'Gladdis',
                config: undefined,
                model: 'whisper-1',
                liveSuffix: 'dictated, but not read',
                readSuffix: 'transcribed and read',
                temperature: 0,
                echoOutput: true,
                deleteFile: false,
                language: undefined,
            },
        })
    })
})
