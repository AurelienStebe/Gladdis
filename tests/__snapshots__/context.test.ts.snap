// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadContext() function > loads the default values 1`] = `
{
  "file": {
    "date": Any<Date>,
    "disk": Any<Object>,
    "name": "input",
    "path": Any<String>,
    "text": "",
  },
  "gladdis": {
    "config": undefined,
    "freq_penalty": 0,
    "label": "Gladdis",
    "model": "gpt-3.5-turbo",
    "pres_penalty": 0,
    "temperature": 0,
    "top_p_param": 100,
  },
  "user": {
    "data": Any<String>,
    "env": Any<Object>,
    "history": [],
    "label": "User",
    "prompt": "",
  },
  "whisper": {
    "config": undefined,
    "deleteFile": false,
    "echoOutput": true,
    "input": "Gladdis",
    "language": undefined,
    "liveSuffix": "dictated, but not read",
    "model": "whisper-1",
    "readSuffix": "transcribed and read",
    "temperature": 0,
  },
}
`;
